src/lib.rs

```diff
@@ -1,20 +1,152 @@
+//! should already help with the vast majority of the cases and we're sufficiently confident that
+//! these errors won't have a breaking change any time soon (possibly never).
+//!
+//! If you're writing a binary you don't need to worry about any of this and just use the unstable
+//! version for now. If you're writing a library you should use these stable errors in the API but
+//! you may internally depend on the unstable crate version to get the advanced features that won't
+//! affect your API. This way your API can stabilize before all features in this crate are fully
+//! stable and you still can use all of them.
+//!
+//! ## Crate features
+//!
+//! * `std` - enables the standard library, on by default.
+//! * `alloc` - enables features that require allocation such as decoding into `Vec<u8>`, implied
+//! by `std`.
+//! * `newer-rust-version` - enables Rust version detection and thus newer features, may add
+//!                          dependency on a feature detection crate to reduce compile times. This
+//!                          feature is expected to do nothing once the native detection is in Rust
+//!                          and our MSRV is at least that version. We may also remove the feature
+//!                          gate in 2.0 with semver trick once that happens.
+//!
+//! ## MSRV policy
+//!
+//! The MSRV of the crate is currently 1.63.0 and we don't intend to bump it until the newer Rust
+//! version is at least two years old and also included in Debian stable (1.63 is in Debian 12 at
+//! the moment).
+//!
+//! Note though that the dependencies may have looser policy. This is not considered breaking/wrong
+//! - you would just need to pin them in `Cargo.lock` (not `.toml`).
 
-#![cfg_attr(all(not(test), not(feature = "std")), no_std)]
+#![no_std]
 // Experimental features we need.
 #![cfg_attr(docsrs, feature(doc_cfg))]
 #![cfg_attr(docsrs, feature(doc_auto_cfg))]
 // Coding conventions
 #![warn(missing_docs)]
 
+#[cfg(feature = "std")]
+extern crate std;
+
 #[cfg(feature = "alloc")]
 extern crate alloc;
+
+mod error;
+mod iter;
+
+#[cfg(feature = "alloc")]
+use alloc::vec::Vec;
+
+use crate::iter::HexToBytesIter;
+
+#[rustfmt::skip]                // Keep public re-exports separate.
+#[doc(inline)]
+pub use self::error::{DecodeFixedSizedBytesError, InvalidCharError, InvalidLengthError};
+#[cfg(feature = "alloc")]
+pub use self::error::{DecodeDynSizedBytesError, OddLengthStringError};
+
+/// Decodes a hex string with variable length.
+///
+/// The length of the returned `Vec` is determined by the length of the input, meaning all even
+/// lengths of the input string are allowed. If you know the required length at compile time using
+/// [`decode_fixed_sized`] is most likely a better choice.
+///
+/// # Errors
+///
+/// Errors if `hex` contains invalid characters or doesn't have even length.
+#[cfg(feature = "alloc")]
+pub fn decode_dyn_sized(hex: &str) -> Result<Vec<u8>, DecodeDynSizedBytesError> {
```

Comment:
In 040a20e1787922dbe483800406ad3c81ef07de82:

Both `dyn` and `sized` mean things in Rust that have nothing to do with this funtion. Calling this `decode_vec` was fine. You could also call it `decode` because the obvious thing that a hex string decodes to is a byte vector.

Reply:
Yeah, I'm not happy with that either. But is there a better way in English to express "this doesn't have fixed length"?

Reply:
A `&str` does not have a fixed length. Nor does a `Vec<u8>`. We don't need to express this. It's in the type signature, and the type signature is the most natural thing that any function called `hex::decode` could have.

If you don't like `decode_vec` I think we should go with `decode`.

Reply:
I'd rather go with `_vec` suffix to not make vecs more convenient than arrays.